diff --git a/NEWS b/NEWS
index 25334b2..a4df18d 100644
--- a/NEWS
+++ b/NEWS
@@ -410,13 +410,6 @@ GNU grep NEWS                                    -*- outline -*-
 
 ** New features
 
-  grep now diagnoses (and fails with exit status 2) commonly mistyped
-  regular expression like [:space:], [:digit:], etc.  Before, those were
-  silently interpreted as [ac:eps] and [dgit:] respectively.  Virtually
-  all who make that class of mistake should have used [[:space:]] or
-  [[:digit:]].  This new behavior is disabled when the POSIXLY_CORRECT
-  environment variable is set.
-
   On systems using glibc, grep can support equivalence classes.  However,
   whether they actually work depends on glibc's locale definitions.
 
diff --git a/doc/grep.texi b/doc/grep.texi
index c9417e5..c05f916 100644
--- a/doc/grep.texi
+++ b/doc/grep.texi
@@ -1000,8 +1000,6 @@ follow file names must be treated as file names;
 by default,
 such options are permuted to the front of the operand list
 and are treated as options.
-Also, @code{POSIXLY_CORRECT} disables special handling of an
-invalid bracket expression.  @xref{invalid-bracket-expr}.
 
 @item _@var{N}_GNU_nonoption_argv_flags_
 @vindex _@var{N}_GNU_nonoption_argv_flags_ @r{environment variable}
@@ -1333,13 +1331,6 @@ Note that the brackets in these class names are
 part of the symbolic names, and must be included in addition to
 the brackets delimiting the bracket expression.
 
-@anchor{invalid-bracket-expr}
-If you mistakenly omit the outer brackets, and search for say, @samp{[:upper:]},
-GNU @command{grep} prints a diagnostic and exits with status 2, on
-the assumption that you did not intend to search for the nominally
-equivalent regular expression: @samp{[:epru]}.
-Set the @code{POSIXLY_CORRECT} environment variable to disable this feature.
-
 Most meta-characters lose their special meaning inside bracket expressions.
 
 @table @samp
diff --git a/src/dfa.c b/src/dfa.c
index 48a83cd..50a0c86 100644
--- a/src/dfa.c
+++ b/src/dfa.c
@@ -1213,9 +1213,6 @@ parse_bracket_exp (void)
     }
   while ((wc = wc1, (c = c1) != ']'));
 
-  if (colon_warning_state == 7)
-    dfawarn (_("character class syntax is [[:space:]], not [:space:]"));
-
   if (! known_bracket_exp)
     return BACKREF;
 
diff --git a/tests/Makefile.am b/tests/Makefile.am
index 31e2a81..5282d35 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -107,7 +107,6 @@ TESTS =						\
   turkish-eyes					\
   turkish-I					\
   turkish-I-without-dot				\
-  warn-char-classes				\
   word-delim-multibyte				\
   word-multi-file				\
   yesno
diff --git a/tests/Makefile.in b/tests/Makefile.in
index e42862d..6549f66 100644
--- a/tests/Makefile.in
+++ b/tests/Makefile.in
@@ -1406,7 +1406,6 @@ TESTS = \
   turkish-eyes					\
   turkish-I					\
   turkish-I-without-dot				\
-  warn-char-classes				\
   word-delim-multibyte				\
   word-multi-file				\
   yesno
@@ -2265,13 +2264,6 @@ turkish-I-without-dot.log: turkish-I-without-dot
 	--log-file $$b.log --trs-file $$b.trs \
 	$(am__common_driver_flags) $(AM_LOG_DRIVER_FLAGS) $(LOG_DRIVER_FLAGS) -- $(LOG_COMPILE) \
 	"$$tst" $(AM_TESTS_FD_REDIRECT)
-warn-char-classes.log: warn-char-classes
-	@p='warn-char-classes'; \
-	b='warn-char-classes'; \
-	$(am__check_pre) $(LOG_DRIVER) --test-name "$$f" \
-	--log-file $$b.log --trs-file $$b.trs \
-	$(am__common_driver_flags) $(AM_LOG_DRIVER_FLAGS) $(LOG_DRIVER_FLAGS) -- $(LOG_COMPILE) \
-	"$$tst" $(AM_TESTS_FD_REDIRECT)
 word-delim-multibyte.log: word-delim-multibyte
 	@p='word-delim-multibyte'; \
 	b='word-delim-multibyte'; \
