diff -upr a/gnulib-tests/test-update-copyright.sh b/gnulib-tests/test-update-copyright.sh
--- a/gnulib-tests/test-update-copyright.sh	2014-01-02 02:32:07.000000000 +0100
+++ b/gnulib-tests/test-update-copyright.sh	2014-02-04 11:49:27.000000000 +0100
@@ -228,7 +228,7 @@ Copyright (C) 1990-2005, 2007-2009 Acme,
 # Software Foundation, Inc.
 EOF
 
-rm $TMP*
+rm -f $TMP*
 
 ## -------------- ##
 ## Current year.  ##
@@ -254,7 +254,7 @@ compare /dev/null $TMP-stderr || exit 1
 compare - $TMP <<EOF || exit 1
 '\" Copyright (C) 2006, $YEAR Free Software Foundation, Inc.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## ------------------ ##
 ## Surrounding text.  ##
@@ -278,7 +278,7 @@ dnl Undisturbed text.
 dnl Copyright (C) 1989, 2010 Free Software Foundation, Inc.
 dnl   Undisturbed text.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## --------------- ##
 ## Widest prefix.  ##
@@ -307,7 +307,7 @@ compare - $TMP <<EOF || exit 1
 #### Copyright (C) 1976-1988, 1999-2008, 2010-2011 Free Software
 #### Foundation, Inc.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## ------------------- ##
 ## Prefix too large.  ##
@@ -330,7 +330,7 @@ compare - $TMP <<EOF || exit 1
 ####  1986, 1987, 1988, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007,
 ####  2008 Free Software Foundation, Inc.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## ------------- ##
 ## Blank lines.  ##
@@ -365,7 +365,7 @@ Copyright (C) 1976, 1977, 1978, 1979, 19
 1986, 1987, 1988, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007,
 2008 Free Software Foundation, Inc.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## -------------- ##
 ## Leading tabs.  ##
@@ -394,7 +394,7 @@ compare - $TMP <<EOF || exit 1
 	Copyright (C) 1987-1988, 1991-2011 Free Software Foundation,
 	Inc.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## -------------------- ##
 ## Unusual whitespace.  ##
@@ -431,7 +431,7 @@ compare /dev/null $TMP-stderr || exit 1
 compare - $TMP <<EOF || exit 1
 		# Copyright (C) 1987-2011 Free Software Foundation, Inc.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## --------- ##
 ## DOS EOL.  ##
@@ -453,7 +453,7 @@ Rem 1998, 1999, 2000, 2001, 2002, 2003,
 Rem 2009, 2010 Free Software Foundation, Inc.@
 EOF
 compare $TMP-exp $TMP || exit 1
-rm $TMP*
+rm -f $TMP*
 
 ## --------------- ##
 ## Omitted "(C)".  ##
@@ -474,7 +474,7 @@ compare - $TMP <<EOF || exit 1
   1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010
   Free Software Foundation, Inc.
 EOF
-rm $TMP*
+rm -f $TMP*
 
 ## ------------------ ##
 ## C-style comments.  ##
@@ -542,6 +542,6 @@ EOF
 compare - $TMP.two-digit-final-is-substr-of-first <<EOF || exit 1
  /* Copyright 1991, 1999, 2010 Free Software Foundation, Inc. */
 EOF
-rm $TMP*
+rm -f $TMP*
 
 exit 0
diff -upr grep-2.16/gnulib-tests/test-version-etc.sh grep-2.16_fixed/gnulib-tests/test-version-etc.sh
--- a/gnulib-tests/test-version-etc.sh	2014-01-02 02:32:07.000000000 +0100
+++ b/gnulib-tests/test-version-etc.sh	2014-02-04 11:50:00.000000000 +0100
@@ -38,6 +38,6 @@ EOT
  tr -d '\015' |
  diff -c $TMP - || ERR=1
 
-rm $TMP
+rm -f $TMP
 
 exit $ERR
