diff --git a/src/Makefile.am b/src/Makefile.am
index e2c82a4..a31f45d 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -21,8 +21,7 @@ AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS)
 # Tell the linker to omit references to unused shared libraries.
 AM_LDFLAGS = $(IGNORE_UNUSED_LIBRARIES_CFLAGS)
 
-bin_PROGRAMS = grep
-bin_SCRIPTS = egrep fgrep
+bin_PROGRAMS = grep
 grep_SOURCES = grep.c searchutils.c \
           dfa.c dfasearch.c \
           kwset.c kwsearch.c \
@@ -42,22 +41,9 @@ grep_LDADD = $(LDADD) $(LIB_PCRE)
 localedir = $(datadir)/locale
 AM_CPPFLAGS = -I$(top_builddir)/lib -I$(top_srcdir)/lib
 
-EXTRA_DIST = dosbuf.c egrep.sh
-
-egrep fgrep: egrep.sh Makefile
-	$(AM_V_GEN)grep=`echo grep | sed -e '$(transform)'`	&& \
-	case $@ in egrep) option=-E;; fgrep) option=-F;; esac	&& \
-	shell_does_substrings='set x/y && d=$${1%/*} && test "$$d" = x' && \
-	if $(SHELL) -c "$$shell_does_substrings" 2>/dev/null; then \
-	  edit_substring='s,X,X,'; \
-	else \
-	  edit_substring='s,\$${0%/\*},`expr "X$$0" : '\''X\\(.*\\)/'\''`,g'; \
-	fi && \
-	sed -e 's|[@]SHELL@|$(SHELL)|g' \
-	    -e "$$edit_substring" \
-	    -e "s|[@]grep@|$$grep|g" \
-	    -e "s|[@]option@|$$option|g" <$(srcdir)/egrep.sh >$@-t
-	$(AM_V_at)chmod +x $@-t
-	$(AM_V_at)mv $@-t $@
+EXTRA_DIST = dosbuf.c
+
+egrep fgrep: Makefile
+	$(LN) -s grep $@
 
 CLEANFILES = egrep fgrep *-t
diff --git a/src/Makefile.in b/src/Makefile.in
index 2cc8955..4600519 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -94,7 +94,7 @@ PRE_UNINSTALL = :
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-bin_PROGRAMS = grep$(EXEEXT)
+bin_PROGRAMS = grep$(EXEEXT)
 subdir = src
 DIST_COMMON = $(srcdir)/Makefile.in $(srcdir)/Makefile.am \
 	$(top_srcdir)/build-aux/depcomp $(noinst_HEADERS)
@@ -1155,7 +1155,6 @@ AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS)
 
 # Tell the linker to omit references to unused shared libraries.
 AM_LDFLAGS = $(IGNORE_UNUSED_LIBRARIES_CFLAGS)
-bin_SCRIPTS = egrep fgrep
 grep_SOURCES = grep.c searchutils.c \
           dfa.c dfasearch.c \
           kwset.c kwsearch.c \
@@ -1174,7 +1173,7 @@ LDADD = \
 
 grep_LDADD = $(LDADD) $(LIB_PCRE)
 AM_CPPFLAGS = -I$(top_builddir)/lib -I$(top_srcdir)/lib
-EXTRA_DIST = dosbuf.c egrep.sh
+EXTRA_DIST = dosbuf.c
 CLEANFILES = egrep fgrep *-t
 all: all-am
 
@@ -1529,21 +1528,8 @@ uninstall-am: uninstall-binPROGRAMS uninstall-binSCRIPTS
 .PRECIOUS: Makefile
 
 
-egrep fgrep: egrep.sh Makefile
-	$(AM_V_GEN)grep=`echo grep | sed -e '$(transform)'`	&& \
-	case $@ in egrep) option=-E;; fgrep) option=-F;; esac	&& \
-	shell_does_substrings='set x/y && d=$${1%/*} && test "$$d" = x' && \
-	if $(SHELL) -c "$$shell_does_substrings" 2>/dev/null; then \
-	  edit_substring='s,X,X,'; \
-	else \
-	  edit_substring='s,\$${0%/\*},`expr "X$$0" : '\''X\\(.*\\)/'\''`,g'; \
-	fi && \
-	sed -e 's|[@]SHELL@|$(SHELL)|g' \
-	    -e "$$edit_substring" \
-	    -e "s|[@]grep@|$$grep|g" \
-	    -e "s|[@]option@|$$option|g" <$(srcdir)/egrep.sh >$@-t
-	$(AM_V_at)chmod +x $@-t
-	$(AM_V_at)mv $@-t $@
+egrep fgrep: Makefile
+	$(LN) -s grep $@
 
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
diff --git a/src/egrep.sh b/src/egrep.sh
deleted file mode 100644
index 1a03d2a..0000000
--- a/src/egrep.sh
+++ /dev/null
@@ -1,11 +0,0 @@
-#!@SHELL@
-grep=grep
-case $0 in
-  */*)
-    dir=${0%/*}
-    if test -x "$dir/@grep@"; then
-      PATH=$dir:$PATH
-      grep=@grep@
-    fi;;
-esac
-exec $grep @option@ "$@"
diff --git a/src/grep.c b/src/grep.c
index 2208a4e..6fe6ac9 100644
--- a/src/grep.c
+++ b/src/grep.c
@@ -26,6 +26,7 @@
 #include <fcntl.h>
 #include <inttypes.h>
 #include <stdio.h>
+#include <string.h>
 #include "system.h"
 
 #include "argmatch.h"
@@ -1938,6 +1939,7 @@ fgrep_to_grep_pattern (size_t len, char const *keys,
 int
 main (int argc, char **argv)
 {
+  char *ptr_c;
   char *keys;
   size_t keycc, oldcc, keyalloc;
   int with_filenames;
@@ -1984,6 +1986,19 @@ main (int argc, char **argv)
   compile = matchers[0].compile;
   execute = matchers[0].execute;
 
+  /* Check if executed through symlinks egrep/fgrep */
+  if (argv[0] != NULL)
+    {
+      ptr_c = strchrnul(argv[0], ' ');
+      if (ptr_c - argv[0] >= 5)
+        {
+          if (strncmp(ptr_c - 5, "egrep", 5) == 0)
+            setmatcher ("egrep");
+          else if (strncmp(ptr_c - 5, "fgrep", 5) == 0)
+            setmatcher ("fgrep");
+        }
+    }
+
   while (prev_optind = optind,
          (opt = get_nondigit_option (argc, argv, &default_context)) != -1)
     switch (opt)
